stablehlo/reference/ProcessGrid.h

```diff
@@ -0,0 +1,127 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_REFERENCE_PROCESSGRID_H
+#define STABLEHLO_REFERENCE_PROCESSGRID_H
+
+#include <condition_variable>
+#include <cstdint>
+#include <map>
+#include <mutex>
+#include <utility>
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/reference/Tensor.h"
+
+namespace mlir {
+namespace stablehlo {
+
+using ChannelId = int64_t;
+
+// StableHLO `process_id`.
+struct ProcessId {
+  /// StableHLO `replica_id`.
+  uint32_t replicaId;
+
+  /// StableHLO `partition_id`.
+  uint32_t partitionId;
+
+  bool operator<(const ProcessId &other) const;
```

Comment:
It is still needed to sort the keys of `std::map<std::pair<ProcessGroup, ChannelId>, RendezvousResult> channels_`. We could use `std::unordered_map`, but this requires a hash function instead. We will soon need to use this operator to sort `RendezvousResult` anyways hence keeping this change.

Reply:
"We will soon need to use this operator to sort RendezvousResult anyways hence keeping this change". I think this operator should come with an update to the spec, because this piece is underspecced at the moment. Let's defer to a later PR.

"It is still needed to sort the keys". Hmm good point. Let's switch to a different container then. Let's use llvm::DenseMap to follow LLVM conventions.

Reply:
"because this piece is underspecced at the moment". Actually, I was wrong - the spec does talk about how `process_groups` are computed exactly. But I still think we shouldn't have `operator<`, since this can be an implementation detail of `RendezvousResult`.

Reply:
>  Let's switch to a different container then. Let's use llvm::DenseMap to follow LLVM conventions.

I looked into `std::map`, `std::unordered_map`, and `llvm::DenseMap`, and of the three, `std::map` seemed to have the most straightforward, minimal implementation. Since we define how `process_groups` are computed for each communication strategy, we could define `operator<` on the `ProcessGroup`, but this seems like an anti-pattern. I've added a comment to justify the need.

Reply:
I opened a [ticket](https://github.com/openxla/stablehlo/issues/1720) to address this in the future.